src/source/blockedsource.js

```diff
@@ -82,92 +84,66 @@ export class BlockedSource extends BaseSource {
 
       const firstBlockOffset = Math.floor(offset / this.blockSize) * this.blockSize;
 
-      // chunk the current slice into blocks
       for (let current = firstBlockOffset; current < top; current += this.blockSize) {
-        // check if the block is cached, being requested or still missing
         const blockId = Math.floor(current / this.blockSize);
-
-        if (this.blockCache.has(blockId)) {
-          cachedBlocks.set(blockId, this.blockCache.get(blockId));
-        } else if (this.blockRequests.has(blockId)) {
-          blockRequests.set(blockId, this.blockRequests.get(blockId));
-        } else if (this.blockIdsToFetch.has(blockId)) {
-          missingBlockIds.add(blockId);
-        } else {
+        if (!this.blockCache.has(blockId) && !this.blockRequests.has(blockId)) {
           this.blockIdsToFetch.add(blockId);
-          missingBlockIds.add(blockId);
+          missingBlockIds.push(blockId);
         }
+        if (this.blockRequests.has(blockId)) {
+          blockRequests.push(this.blockRequests.get(blockId));
+        }
+        allBlockIds.push(blockId);
       }
     }
 
     // allow additional block requests to accumulate
     await wait();
     this.fetchBlocks(signal);
 
+    // Gather all of the new requests that this fetch call is contributing to `fetch`.
+    const missingRequests = [];
     for (const blockId of missingBlockIds) {
-      const block = this.blockRequests.get(blockId);
-      const cachedBlock = this.blockCache.get(blockId);
-
-      if (block) {
-        blockRequests.set(blockId, block);
-      } else if (cachedBlock) {
-        cachedBlocks.set(blockId, cachedBlock);
-      } else {
-        throw new Error(`Block ${blockId} is not in the block requests`);
+      if (this.blockRequests.has(blockId)) {
+        missingRequests.push(this.blockRequests.get(blockId));
       }
```

Comment:
I think just throwing an Error is fine.  I'll update.

Reply:
Oh this condition does have a purpose.  Once a request is completed and its data added to the catch, `this.blockRequests` no longer contains the block.   But that is ok because it is in the cache so I think leaving this as-is is ok.  I'll just add a comment.